### PR TITLE
add reduce_split pass

### PR DESCRIPTION
**Pull request by zhiqiu**
This PR tries to optimize the performance of reduce_xx by split it into two kernels when the reduce_numel is large.

For example, 
Given a with shape [64,14,14,256]
b = ReduceSum(a, axis=0,1,2)

The reduce_numel is 64*14*14=12544.

The original kernel generated uses 1 block and 1024 thread, so each thread reduces 64*14*14*256/1024=3136 items, which is slow in cuda.

Now, it first reshapes a as [112, 112, 256], reduce it into [112, 256], and then futher reduce it into [256].
The first kernel uses 1 block and 1024 threads, so each thread reduces 28 items.
The second kernel uses 1 block and 1024 threads, so each thread reduces 28 items.

The nsys shows the performance of 3 kernels as below:

- original: fn_reduce_sum0_kernel,  2127us
- now: fn_reshape_reduce_sum_6_kernel + fn_reduce_sum_0_reshape_1_5_kernel,   78us

image

### CODE REVIEW

File: cinn/runtime/flags.cc [hunk 1]
```diff
@@ -70,6 +70,8 @@ DEFINE_bool(cinn_ir_schedule,
             BoolFromEnv("FLAGS_cinn_ir_schedule", true),
             "Whether use reconstructed schedule primitives.");
 
+DEFINE_bool(use_reduce_split_pass, BoolFromEnv("FLAGS_use_reduce_split_pass", false), "Whether use reduce split pass.");
```
Comment (SunNy820828449) [inline cinn/runtime/flags.cc:73]: 直接默认打开就行 不用加一个flag

File: cinn/hlir/pass/reduce_split_pass.cc [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+}  // namespace pass
+}  // namespace hlir
+}  // namespace cinn
+
+CINN_REGISTER_HELPER(ReduceSplit) {
+  CINN_REGISTER_PASS(ReduceSplit)
+      .describe("")
+      .set_change_structure(false)
```
Comment (SunNy820828449) [inline cinn/hlir/pass/reduce_split_pass.cc]: change structure = true

File: cinn/frontend/optimize.cc [hunk 3]
```diff
@@ -74,6 +75,11 @@ OptimizeOptions DefaultTrainingOptimizeOptions() {
     options.graph_passes.emplace_back("CommonSubexpressionEliminationPass");
   }
 
+  // this pass should be applied before merge
+  if (FLAGS_use_reduce_split_pass) {
+    options.graph_passes.emplace_back("ReduceSplit");
```
Comment (SunNy820828449) [inline cinn/frontend/optimize.cc:80]: 默认打开就行 直接使用这个pass

File: cinn/hlir/pass/reduce_split_pass.cc [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  static std::unordered_set<std::string> reduce_op_type = {
+      "reduce_sum", "reduce_mean", "reduce_max", "reduce_min", "reduce_all", "reduce_any"};
+  if (reduce_op_type.count(node->op()->name)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+std::pair<int, int> DivideToClosetNum(int n) {
```
Comment (SunNy820828449) [inline cinn/hlir/pass/reduce_split_pass.cc:41]: 有个问题：把n分成a * b好还是b * a好？其中a > b

File: cinn/hlir/pass/reduce_split_pass.cc [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        auto reshape0_data = new NodeData(Shared<Node>(reshape0), 0, 0, common::UniqName("reshape"), false);
+        graph->RegisterNode(reshape0_data->id(), reshape0_data);
+        reshape0->LinkTo(reshape0_data);
+        shape_dict[reshape0_data->id()] = absl::get<std::vector<int>>(reshape0->attrs.attr_store.at("shape"));
+        dtype_dict[reshape0_data->id()] = common::Str2Type(common::Type2Str(dtype_dict[in->id()]));
+
+        // create reduce node0
+        Node* reduce0                         = new Node(Operator::Get(name), name, common::UniqName(name));
+        reduce0->attrs.attr_store["dim"]      = std::vector<int>{0};
+        reduce0->attrs.attr_store["keep_dim"] = absl::get<bool>(n->attrs.attr_store.at("keep_dim"));
```
Comment (SunNy820828449) [inline cinn/hlir/pass/reduce_split_pass.cc:135]: 第一个 keep_dim = false, 输出shape b * c
第二个 keep_dim = false  输出shape c
如果 原来的reduce keep_dim = true, 加一个reshape为[1,1,1,1,c]

File: cinn/hlir/pass/reduce_split_pass.cc [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+                                                  : std::vector<int>{reduce_numel1, in_shape[in_shape.size() - 1]};
+        dtype_dict[reduce0_data->id()] = common::Str2Type(common::Type2Str(dtype_dict[in->id()]));
+
+        // create reduce node1
+        Node* reduce1                         = new Node(Operator::Get(name), name, common::UniqName(name));
+        reduce1->attrs.attr_store["dim"]      = keep_dim ? std::vector<int>{0, 1} : std::vector<int>{0};
+        reduce1->attrs.attr_store["keep_dim"] = absl::get<bool>(n->attrs.attr_store.at("keep_dim"));
+        graph->RegisterNode(reduce1->id(), reduce1);
+        reduce0_data->LinkTo(reduce1);
+        auto reduce1_data = new NodeData(Shared<Node>(reduce1), 0, 0, common::UniqName("reduce"), false);
```
Comment (SunNy820828449) [inline cinn/hlir/pass/reduce_split_pass.cc]: NodeData名字使用UniqName("var")

File: cinn/hlir/pass/reduce_split_pass.cc [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        shape_dict[reshape0_data->id()] = absl::get<std::vector<int>>(reshape0->attrs.attr_store.at("shape"));
+        dtype_dict[reshape0_data->id()] = common::Str2Type(common::Type2Str(dtype_dict[in->id()]));
+
+        // create reduce node0
+        Node* reduce0                         = new Node(Operator::Get(name), name, common::UniqName(name));
+        reduce0->attrs.attr_store["dim"]      = std::vector<int>{0};
+        reduce0->attrs.attr_store["keep_dim"] = absl::get<bool>(n->attrs.attr_store.at("keep_dim"));
+        graph->RegisterNode(reduce0->id(), reduce0);
+        reshape0_data->LinkTo(reduce0);
+        auto reduce0_data = new NodeData(Shared<Node>(reduce0), 0, 0, common::UniqName("reduce"), false);
```
Comment (SunNy820828449) [inline cinn/hlir/pass/reduce_split_pass.cc]: NodeData名字使用UniqName("var")

File: cinn/hlir/pass/reduce_split_pass.cc [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        if ((!all_preceding_dim_reduced) || numel <= MAX_NUM_THREADS * MAX_ITER_PER_THREAD) {
+          continue;
+        }
+
+        auto res          = DivideToClosetNum(reduce_numel);
+        int reduce_numel0 = std::get<0>(res), reduce_numel1 = std::get<1>(res);
+        VLOG(4) << "reduce_numel0: " << reduce_numel0 << " reduce_numel1: " << reduce_numel1;
+
+        // create reshape node0
+        Node* reshape0 = new Node(Operator::Get("reshape"), "reshape", common::UniqName("reshape"));
```
Comment (SunNy820828449) [inline cinn/hlir/pass/reduce_split_pass.cc]: 如果输入的shape本来就是[N,C] 第一个reshape可以不加

File: cinn/hlir/pass/reduce_split_pass.cc (hunk 7 above)
Comment (zhiqiu) [reply cinn/hlir/pass/reduce_split_pass.cc]: 好的

File: cinn/hlir/pass/reduce_split_pass.cc (hunk 4 above)
Comment (zhiqiu) [reply cinn/hlir/pass/reduce_split_pass.cc:41]: 好问题，我理解差别很小，因为这里逻辑a和b是非常接近的。

File: cinn/hlir/pass/reduce_split_pass.cc (hunk 5 above)
Comment (zhiqiu) [reply cinn/hlir/pass/reduce_split_pass.cc:135]: 是的，只是我觉得多余的reshape最后会被fuse掉，所以先统一加了。

File: cinn/hlir/pass/reduce_split_pass.cc (hunk 8 above)
Comment (zhiqiu) [reply cinn/hlir/pass/reduce_split_pass.cc]: 同上

File: cinn/hlir/pass/reduce_split_pass.cc (hunk 2 above)
Comment (zhiqiu) [reply cinn/hlir/pass/reduce_split_pass.cc]: done

File: cinn/runtime/flags.cc (hunk 1 above)
Comment (zhiqiu) [reply cinn/runtime/flags.cc:73]: 当前还没有在很多模型中验证效果，所以先加个flag，后面可以再去掉。

**Review by SunNy820828449 — APPROVED** [review]
LGTM

**Pull request merged**